aiohttp/connector.py

```diff
@@ -477,35 +490,37 @@ def _get(self, key):
         del self._conns[key]
         return None
 
-    def _release_key_waiter(self, key, waiters):
-        if not waiters:
-            return False
+    def _release_waiter(self):
+        """
+        Iterates over all waiters till found one that is not finsihed and
+        belongs to a host that has available connections.
+        """
+        if not self._waiters:
+            return
 
-        waiter = waiters.popleft()
-        if not waiter.done():
-            waiter.set_result(None)
+        # Having the dict keys ordered this avoids to iterate
+        # at the same order at each call.
+        queues = list(self._waiters.keys())
+        random.shuffle(queues)
 
-        if not waiters:
-            del self._waiters[key]
+        for key in queues:
+            if self._available_connections(key) < 1:
+                continue
 
-        return True
+            waiters = self._waiters[key]
 
-    def _release_waiter(self):
-        # always release only one waiter
+            while waiters:
+                waiter = waiters.popleft()
+                if not waiter.done():
+                    waiter.set_result(None)
 
-        if self._limit:
-            # if we have limit and we have available
-            if self._limit - len(self._acquired) > 0:
-                for key, waiters in self._waiters.items():
-                    if self._release_key_waiter(key, waiters):
-                        break
-
-        elif self._limit_per_host:
-            # if we have dont have limit but have limit per host
-            # then release first available
-            for key, waiters in self._waiters.items():
-                if self._release_key_waiter(key, waiters):
-                    break
+                    if not waiters:
+                        del self._waiters[key]
+
+                    return
+
+            if not waiters:
```

Comment:
The condition is always true.

Reply:
Yes it doesn't make sense right now let me get rid of this part of the code and I will check the tests.